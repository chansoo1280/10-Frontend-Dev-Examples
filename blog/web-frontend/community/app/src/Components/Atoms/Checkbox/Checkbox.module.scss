@import "@Styles/scss/default.scss";
@keyframes ani-check {
    0% {
        transform:scale(0.5) translateY(5px);
        opacity:0;
    }
    80%{
        transform:scale(1.1) translateY(0);
        opacity:1;
    }
    100%{
        transform:scale(1);
    }
}
.checkbox{
    display:flex;
    align-items:center;
    gap:8px;
    &__checkbox{
        cursor: pointer;
        position:relative;
        margin:0;
        padding:0;
        width:16px;
        height:16px;
        border:1px solid;
        outline:none;
        border-radius:$border-radius;
        transition:background-color .2s ease, border-color .2s ease, color .2s ease;
        @include click-animating(0.5);
        @include set-color("", $white, $gray-5);
        &:checked{
            @include set-color("", $primary-6, $primary-6);
            &:before{
                animation:ani-check .4s cubic-bezier(.08,.82,.17,1);
            }
        }
        &:before{
            content:"";
            display: block;
            width:100%;
            height:100%;
            transition:opacity .24s ease, transform .4s ease;
            transform-origin:center bottom;
            background:url(/Images/icon_check.svg) center no-repeat;
        }
    }
}