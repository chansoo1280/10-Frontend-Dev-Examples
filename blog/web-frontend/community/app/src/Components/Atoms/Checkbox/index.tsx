// #region Global Imports
import React, { ChangeEvent, ChangeEventHandler, useRef, useState } from "react"
import classNames from "classnames"
// #endregion Global Imports

// #region Local Imports
import { defaultProps } from "@Components"
import Typography from "../Typography"
import { useClickAnimating } from "@Hooks"
import styles from "./Checkbox.module.scss"
// #endregion Local Imports
const { Text } = Typography
interface CheckboxProps extends defaultProps {
    checked?: boolean
    onChange: ChangeEventHandler<HTMLInputElement>
    label?: string
}

const Checkbox = (props: CheckboxProps): JSX.Element => {
    const { checked = false, id, show, className, onChange, disabled, label, ...rest } = props
    const checkboxRef = useRef<HTMLInputElement>(null)
    const [value, setValue] = useState(checked)
    const { isClick, execClickAnimation } = useClickAnimating()
    const prefixCls = "checkbox"
    const classes = classNames(
        styles[`${prefixCls}`],
        {
            [styles[`${prefixCls}--hide`]]: show === false,
        },
        className,
    )
    const handleChange = (e: ChangeEvent<HTMLInputElement>) => {
        if (disabled) {
            return
        }
        if (e.currentTarget.checked === true) {
            execClickAnimation()
        }
        setValue(!value)
        onChange(e)
    }
    const boxProps = {
        className: classes,
        ...rest,
    }
    const checkboxProps = {
        ref: checkboxRef,
        id,
        className: classNames(styles[`${prefixCls}__checkbox`]),
        onChange: handleChange,
        disabled,
        checked: value,
        "data-click-animating": isClick === true,
    }
    return (
        <div {...boxProps}>
            <input type="checkbox" {...checkboxProps} />
            {label && (
                <label htmlFor={id}>
                    <Text size="small">{label}</Text>
                </label>
            )}
        </div>
    )
}
export default Checkbox
