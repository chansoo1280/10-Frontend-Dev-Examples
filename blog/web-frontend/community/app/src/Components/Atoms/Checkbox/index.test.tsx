import { fireEvent, render, screen } from "@testing-library/react"
import Checkbox from "."
describe("Checkbox", () => {
    it("renders checkbox unchanged", () => {
        const handleChange = jest.fn()
        const { container } = render(<Checkbox onChange={handleChange} />)
        expect(container).toMatchSnapshot()
    })
    it("should render a checkbox with the class", () => {
        const handleChange = jest.fn()
        render(<Checkbox onChange={handleChange} data-testid="checkbox" />)

        const checkbox = screen.getByTestId("checkbox")
        expect(checkbox).toBeInTheDocument()
        expect(checkbox).toHaveClass("checkbox")
    })
    it("클릭 후 값 확인", () => {
        const handleChange = jest.fn()
        render(<Checkbox id="checkbox" label="체크박스" onChange={handleChange} data-testid="checkbox" />)

        const checkbox = screen.getByLabelText("체크박스")
        fireEvent.click(checkbox)
        expect(checkbox).toBeChecked()
    })
    it("비활성화 상태에서 클릭 후 값 확인", () => {
        const handleChange = jest.fn()
        render(<Checkbox id="checkbox" label="체크박스" onChange={handleChange} data-testid="checkbox" disabled />)

        const checkbox = screen.getByLabelText("체크박스")
        fireEvent.click(checkbox)
        expect(checkbox).not.toBeChecked()
    })
})
