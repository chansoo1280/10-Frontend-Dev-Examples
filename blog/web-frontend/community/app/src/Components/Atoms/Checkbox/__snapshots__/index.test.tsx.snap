// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Checkbox renders checkbox unchanged 1`] = `
<div>
  <div
    class="checkbox"
  >
    <input
      class="checkbox__checkbox"
      data-click-animating="false"
      type="checkbox"
    />
  </div>
</div>
`;
